streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerWithKey.java

```diff
@@ -75,14 +75,20 @@
     void init(final ProcessorContext context);
 
     /**
-     * Transform the given [key and ]value to a new value.
+     * Transform the given [key and] value to a new value.
      * Additionally, any {@link StateStore} that is {@link KStream#transformValues(ValueTransformerWithKeySupplier, String...)
      * attached} to this operator can be accessed and modified arbitrarily (cf.
      * {@link ProcessorContext#getStateStore(String)}).
      * <p>
-     * Note, that using {@link ProcessorContext#forward(Object, Object)} or
+     * Note that using {@link ProcessorContext#forward(Object, Object)} or
      * {@link ProcessorContext#forward(Object, Object, To)} is not allowed within {@code transform} and
      * will result in an {@link StreamsException exception}.
+     * <p>
+     * Note that if a {@code ValueTransformerWithKey} is used in a {@link KTable#transformValues(ValueTransformerWithKeySupplier, String...)}
+     * (or any other overload of {@code KTable#transformValues(...)}) operation,
+     * then the provided {@link ProcessorContext} from {@link #init(ProcessorContext)}
+     * does not guarantee that all context information will be available when {@code transform()}
+     * is executed.
```

Comment:
We might as well formally specify which things may be missing. Otherwise, it's not clear what users should do to guard their code.
```suggestion
     * does not guarantee that all record context information will be available when
     * {@code transform()} is executed. For example, {@code transform()} may be
     * invoked as a result of upstream punctuations or other out-of-band operations,
     * in which case there will be no headers, topic name, timestamp, or offset available.
     * When those properties are absent, the following placeholder values will
     * be filled in: empty headers, `null` topic name, `-1` timestamp, and `-1` offset.
     * Implementations of this interface should guard against this case.
```

Reply:
Well, for `KTable#transformValues` those things don't apply because there is no out-of-band transformation upstream.

It's the DSL, not the PAPI. It might might sense to add this information in general, but seems to be out-of-scope for this PR?

Reply:
All I was trying to do was characterize why the context information might be missing. If there are _no_ out-of-band operations, then there should never be missing context.